File: framec_tests/python/src/state_context/state_context.py
#Emitted from framec_v0.11.2



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class StateContextSm:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
         # Create and initialize start state compartment.
        
        next_compartment = None
        next_compartment = StateContextSmCompartment('__statecontextsm_state_Init', next_compartment)
        next_compartment.state_vars["w"] = 0
        self.__compartment = next_compartment
        self.__next_compartment = None
        self.return_stack = []
        next_compartment.state_vars["w"] = 0
        
        # Initialize domain
        
        self.tape  = []
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def Start(self,):
        self.return_stack.append(None)
        __e = FrameEvent("Start",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def LogState(self,):
        self.return_stack.append(None)
        __e = FrameEvent("LogState",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def Inc(self,):
        self.return_stack.append(None)
        __e = FrameEvent("Inc",None)
        self.__kernel(__e)
        return self.return_stack.pop(-1)
    
    def Next(self,arg: int):
        parameters = {}
        parameters["arg"] = arg
        self.return_stack.append(None)
        __e = FrameEvent("Next",parameters)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    # ===================== Machine Block =================== #
      #  Change [arg:int]
    
    
    # ----------------------------------------
    # $Init
    
    def __statecontextsm_state_Init(self, __e, compartment):
        if __e._message == ">":
            (compartment.state_vars["w"]) = 3
            self.log_do("w",(compartment.state_vars["w"]))
            return
        elif __e._message == "Inc":
            (compartment.state_vars["w"]) = compartment.state_vars["w"] + 1
            self.log_do("w",(compartment.state_vars["w"]))
            self.return_stack[-1] = (compartment.state_vars["w"])
            return
            
        elif __e._message == "LogState":
            self.log_do("w",(compartment.state_vars["w"]))
            return
        elif __e._message == "Start":
            next_compartment = None
            next_compartment = StateContextSmCompartment('__statecontextsm_state_Foo', next_compartment)
            next_compartment.enter_args["a"] = 3
            next_compartment.enter_args["b"] = compartment.state_vars["w"]
            next_compartment.state_vars["x"] = 0
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $Foo
    
    def __statecontextsm_state_Foo(self, __e, compartment):
        if __e._message == ">":
            self.log_do("a",__e._parameters["a"])
            self.log_do("b",__e._parameters["b"])
            (compartment.state_vars["x"]) = __e._parameters["a"] * __e._parameters["b"]
            self.log_do("x",(compartment.state_vars["x"]))
            return
        elif __e._message == "<":
            self.log_do("c",__e._parameters["c"])
            (compartment.state_vars["x"]) = compartment.state_vars["x"] + __e._parameters["c"]
            self.log_do("x",(compartment.state_vars["x"]))
            return
        elif __e._message == "LogState":
            self.log_do("x",(compartment.state_vars["x"]))
            return
        elif __e._message == "Inc":
            (compartment.state_vars["x"]) = compartment.state_vars["x"] + 1
            self.log_do("x",(compartment.state_vars["x"]))
            self.return_stack[-1] = (compartment.state_vars["x"])
            return
            
        elif __e._message == "Next":
            tmp = __e._parameters["arg"] * 10
            self.__compartment.exit_args["c"] = 10
            next_compartment = None
            next_compartment = StateContextSmCompartment('__statecontextsm_state_Bar', next_compartment)
            next_compartment.enter_args["a"] = tmp
            next_compartment.state_args["y"] = compartment.state_vars["x"]
            next_compartment.state_vars["z"] = 0
            self.__transition(next_compartment)
            return
      #  FIXME: Swapping this to 10 * arg causes a parse error!
      #  |Change| [arg:int]
      #      var tmp = x + arg
      #      -> $Bar(tmp)
      #      ^
    
    
    # ----------------------------------------
    # $Bar
    
    def __statecontextsm_state_Bar(self, __e, compartment):
        if __e._message == ">":
            self.log_do("a",__e._parameters["a"])
            self.log_do("y",(compartment.state_args["y"]))
            (compartment.state_vars["z"]) = __e._parameters["a"] + compartment.state_args["y"]
            self.log_do("z",(compartment.state_vars["z"]))
            return
        elif __e._message == "LogState":
            self.log_do("y",(compartment.state_args["y"]))
            self.log_do("z",(compartment.state_vars["z"]))
            return
        elif __e._message == "Inc":
            (compartment.state_vars["z"]) = compartment.state_vars["z"] + 1
            self.log_do("z",(compartment.state_vars["z"]))
            self.return_stack[-1] = (compartment.state_vars["z"])
            return
            
    
    # ===================== Actions Block =================== #
    
    def log_do(self,name: str,val: int):
        raise NotImplementedError
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__statecontextsm_state_Init':
            self.__statecontextsm_state_Init(__e, self.__compartment)
        elif self.__compartment.state == '__statecontextsm_state_Foo':
            self.__statecontextsm_state_Foo(__e, self.__compartment)
        elif self.__compartment.state == '__statecontextsm_state_Bar':
            self.__statecontextsm_state_Bar(__e, self.__compartment)
        
    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment
    
    def state_info(self):
        return self.__compartment.state
        

# ===================== Compartment =================== #

class StateContextSmCompartment:

    def __init__(self,state,parent_compartment):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment
    

File: framec_tests_old/javascript/src/state_context/state_context.js
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

function FrameEvent(message, parameters) {

    var that = {};
    that._message = message;
    that._parameters = parameters;
    that._return = null;
    return that;
    
}

class StateContextSm {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sInit_;
        this.#compartment = new StateContextSmCompartment(this.#state);
        this.#nextCompartment = null;
        this.#compartment.StateVars["w"] = this.#compartment.StateVars["w"] + 1;
        
        // Initialize domain
        this.tape = [];
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    Start() {
        let e = FrameEvent("Start",null);
        this.#mux(e);
    }
    
    LogState() {
        let e = FrameEvent("LogState",null);
        this.#mux(e);
    }
    
    Inc() {
        let e = FrameEvent("Inc",null);
        this.#mux(e);
        return e._return
    }
    
    Next(arg) {
        let e = FrameEvent("Next",{"arg":arg});
        this.#mux(e);
    }
    
    Change(arg) {
        let e = FrameEvent("Change",{"arg":arg});
        this.#mux(e);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sInit_:
                this.#sInit_(e);
                break;
            case this.#sFoo_:
                this.#sFoo_(e);
                break;
            case this.#sBar_:
                this.#sBar_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sInit_(e) {
        switch (e._message) {
            case ">":
                {
                (this.#compartment.StateVars["w"]) = 3;
                this.log_do("w",(this.#compartment.StateVars["w"]));
                
                return;
                }
                
            case "Inc":
                {
                (this.#compartment.StateVars["w"]) = (this.#compartment.StateVars["w"]) + 1;
                this.log_do("w",(this.#compartment.StateVars["w"]));
                e._return = (this.#compartment.StateVars["w"])
                return;
                }
                
            case "LogState":
                {
                this.log_do("w",(this.#compartment.StateVars["w"]));
                
                return;
                }
                
            case "Start":
                {
                let compartment =  new StateContextSmCompartment(this.#sFoo_);
                
                compartment.EnterArgs["a"] = 3;
                compartment.EnterArgs["b"] = this.#compartment.StateVars["w"];
                compartment.StateVars["x"] = this.#compartment.StateVars["x"] + 1;
                
                this.#transition(compartment);
                
                return;
                }
                
        }
    }
    
    #sFoo_(e) {
        switch (e._message) {
            case ">":
                {
                this.log_do("a",(e._parameters["a"]));
                this.log_do("b",(e._parameters["b"]));
                (this.#compartment.StateVars["x"]) = (e._parameters["a"]) * (e._parameters["b"]);
                this.log_do("x",(this.#compartment.StateVars["x"]));
                
                return;
                }
                
            case "<":
                {
                this.log_do("c",(e._parameters["c"]));
                (this.#compartment.StateVars["x"]) = (this.#compartment.StateVars["x"]) + (e._parameters["c"]);
                this.log_do("x",(this.#compartment.StateVars["x"]));
                
                return;
                }
                
            case "LogState":
                {
                this.log_do("x",(this.#compartment.StateVars["x"]));
                
                return;
                }
                
            case "Inc":
                {
                (this.#compartment.StateVars["x"]) = (this.#compartment.StateVars["x"]) + 1;
                this.log_do("x",(this.#compartment.StateVars["x"]));
                e._return = (this.#compartment.StateVars["x"])
                return;
                }
                
            case "Next":
                {
                let tmp = e._parameters["arg"] * 10;
                this.#compartment.ExitArgs["c"] = 10;
                let compartment =  new StateContextSmCompartment(this.#sBar_);
                
                compartment.EnterArgs["a"] = tmp;
                compartment.StateArgs["y"] = this.#compartment.StateVars["x"];
                compartment.StateVars["z"] = this.#compartment.StateVars["z"] + 1;
                
                this.#transition(compartment);
                
                return;
                }
                
              // FIXME: Swapping this to 10 * arg causes a parse error!
			case "Change":
                {
                let tmp = this.#compartment.StateVars["x"] + e._parameters["arg"];
                
                return;
                }
                
        }
    }  // ->> $Bar(tmp)
	
    
    #sBar_(e) {
        switch (e._message) {
            case ">":
                {
                this.log_do("a",(e._parameters["a"]));
                this.log_do("y",(this.#compartment.StateArgs["y"]));
                (this.#compartment.StateVars["z"]) = (e._parameters["a"]) + (this.#compartment.StateArgs["y"]);
                this.log_do("z",(this.#compartment.StateVars["z"]));
                
                return;
                }
                
            case "LogState":
                {
                this.log_do("y",(this.#compartment.StateArgs["y"]));
                this.log_do("z",(this.#compartment.StateVars["z"]));
                
                return;
                }
                
            case "Inc":
                {
                (this.#compartment.StateVars["z"]) = (this.#compartment.StateVars["z"]) + 1;
                this.log_do("z",(this.#compartment.StateVars["z"]));
                e._return = (this.#compartment.StateVars["z"])
                return;
                }
                
            case "Change":
                {
                let tmp = this.#compartment.StateArgs["y"] + this.#compartment.StateVars["z"] + e._parameters["arg"];
                this.log_do("tmp",tmp);
                
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    log_do(name,val) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    state_info() {
        return this.#compartment.state.name;
    }
    
      // ->> $Init
	
    
};

//=============== Compartment ==============//

class StateContextSmCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


/********************

class StateContextSmController extends StateContextSm {

	constructor() {
	  super()
	}
	log_do(name,val) {}
};

********************/

module.exports = StateContextSm
